File: website/models/helper_grouper.py
from collections import OrderedDict

from tissue_colors import TissueColors

class HelperGrouper:
    def __init__(self, cache, transcript, rows):
        self.transcript = transcript
        self.rows = rows

        self.tissueColors = TissueColors(cache)
        self.byID = {r["id"] : r for r in self.rows}

        for rid, r in self.byID.iteritems():
            r["color"] = self.tissueColors.getTissueColor(r["tissue"])
            r["counts"] = float(r["counts"])

    def getGroupedItems(self, skey):
        return { "byTissue" : self.groupByTissue(skey),
                 "byTissueMax" : self.groupByTissueMax(skey),
                 "byValue" : self.sortByValue(skey)}

    def groupByTissue(self, skey):
        sorter = lambda x: x["tissue"]
        self.rows.sort(key = sorter)

        ret = {}
        for row in self.rows:
            t = row["tissue"]
	    if t not in ret:
	        ret[t] = {"tissue" : t,
                          "color": row["color"],
                          "items": []}
            ret[t]["items"].append(row["id"])

        for k, v in ret.iteritems():
            sorter = lambda rid: (self.byID[rid][skey], self.byID[rid]["tissue"])
            ret[k]["items"].sort(key = sorter, reverse = True)
        return ret

    def groupByTissueMax(self, skey):
        sorter = lambda x: x["tissue"]
        self.rows.sort(key = sorter)

        ret = {}
        for row in self.rows:
            t = row["tissue"]
	    if t not in ret:
	        ret[t] = {"tissue" : t,
                          "color": row["color"],
                          "items": [row["id"]]}
            else:
                if self.byID[ret[t]["items"][0]][skey] < row[skey]:
                    ret[t]["items"] = [row["id"]]

        rows = ret.values()
        sorter = lambda rid: self.byID[rid["items"][0]][skey]
        rows.sort(key = sorter, reverse = True)

        ret = {}
        for idx, row in enumerate(rows):
            t = row["tissue"]
            k = str(idx).zfill(3) + '_' + t
	    ret[k] = row
        return ret

    def sortByValue(self, skey):
        sorter = lambda x: x[skey]
        self.rows.sort(key = sorter, reverse = True)

        ret = OrderedDict()
        for idx, row in enumerate(self.rows):
            t = row["tissue"]
            k = str(idx).zfill(3) + '_' + t
	    ret[k] = {"tissue" : t,
                      "color": row["color"],
                      "items": [row["id"]]}
        return ret

File: graphql/src/helpergrouper.ts
import TissueColors from './tissuecolors';
import { natsorter } from './utils';

export default class HelperGrouper {
    transcript;
    rows: Array<object>;
    byID;

    constructor(transcript, rows) {
        this.transcript = transcript;
        this.rows = rows;
        this.byID = this.rows.reduce((obj, r) => ({ ...obj, [r['id']]: r }), {});

        for (const rid of Object.keys(this.byID)) {
            const r = this.byID[rid];
            r['color'] = TissueColors.getTissueColor(r['tissue']);
            r['counts'] = +r['counts'];
        }
    }

    getGroupedItems(skey) {
        return {
            byTissue: this.groupByTissue(skey),
            byTissueMax: this.groupByTissueMax(skey),
            byValue: this.sortByValue(skey),
        };
    }

    static sortByKey = skey => (a, b) => natsorter(a[skey], b[skey]);
    static sortByTissue = HelperGrouper.sortByKey('tissue');
    static sortByKeyThenTissue = skey => (a, b) =>
        HelperGrouper.sortByKey(skey)(a, b) === 0 ? 0 : HelperGrouper.sortByTissue(a, b);

    groupByTissue(skey) {
        this.rows.sort(HelperGrouper.sortByTissue);

        const ret: any = {};
        for (const row of this.rows) {
            const t = row['tissue'];
            if (!(t in ret)) {
                ret[t] = {
                    name: t,
                    tissue: t,
                    color: row['color'],
                    items: [],
                };
            }
            ret[t]['items'].push(row['id']);
        }

        for (const k of Object.keys(ret)) {
            ret[k]['items'].sort(HelperGrouper.sortByKeyThenTissue(skey)).reverse();
        }
        const sortedkeys = Object.keys(ret).sort(natsorter);
        return sortedkeys.map(tissue => ret[tissue]);
    }

    static pad = (n, width) => {
        n = n + '';
        return n.length >= width ? n : new Array(width - n.length + 1).join('0') + n;
    };

    groupByTissueMax(skey) {
        this.rows.sort(HelperGrouper.sortByTissue);

        const ret: any = {};
        for (const row of this.rows) {
            const t = row['tissue'];
            if (!(t in ret)) {
                ret[t] = {
                    name: t,
                    tissue: t,
                    color: row['color'],
                    items: [row['id']],
                };
            }
            if (this.byID[ret[t]['items'][0]][skey] < row[skey]) {
                ret[t]['items'] = [row['id']];
            }
        }

        const rows = Object.keys(ret).map(k => ret[k]);
        const sorter = (a, b) => natsorter(this.byID[a['items'][0]][skey], this.byID[b['items'][0]][skey]);
        rows.sort(sorter).reverse();

        return rows;
    }

    sortByValue(skey) {
        this.rows.sort(HelperGrouper.sortByKey(skey)).reverse();

        return this.rows.map((row, idx) => ({
            name: idx + '_' + row['tissue'],
            tissue: row['tissue'],
            color: row['color'],
            items: [row['id']],
        }));
    }
}
